build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt

```diff
@@ -375,27 +375,34 @@ class ListExternalizer<T>(elementExternalizer: DataExternalizer<T>) :
 class SetExternalizer<T>(elementExternalizer: DataExternalizer<T>) :
     GenericCollectionExternalizer<T, Set<T>>(elementExternalizer, { size -> LinkedHashSet(size) })
 
-class LinkedHashMapExternalizer<K, V>(
+open class MapExternalizer<K, V, M : Map<K, V>>(
     private val keyExternalizer: DataExternalizer<K>,
-    private val valueExternalizer: DataExternalizer<V>
-) : DataExternalizer<LinkedHashMap<K, V>> {
+    private val valueExternalizer: DataExternalizer<V>,
+    private val newMap: (size: Int) -> MutableMap<K, V> = { size -> LinkedHashMap(size) }
+) : DataExternalizer<M> {
 
-    override fun save(output: DataOutput, map: LinkedHashMap<K, V>) {
+    override fun save(output: DataOutput, map: M) {
         output.writeInt(map.size)
         for ((key, value) in map) {
             keyExternalizer.save(output, key)
             valueExternalizer.save(output, value)
         }
     }
 
-    override fun read(input: DataInput): LinkedHashMap<K, V> {
+    override fun read(input: DataInput): M {
         val size = input.readInt()
-        val map = LinkedHashMap<K, V>(size)
+        val map = newMap(size)
         repeat(size) {
             val key = keyExternalizer.read(input)
             val value = valueExternalizer.read(input)
             map[key] = value
         }
-        return map
+        @Suppress("UNCHECKED_CAST")
```

Comment:
Ivan's comment from https://github.com/google/kotlin/commit/38227b5843d51941f80b8b96585ef74fd363e860#r64563620

> To make it a bit more type-safe, maybe about having default value for `newMap` and change it to `(Int) -> M`?

I just tried it and `map[key] = value` on line 398 above will not compile because `M` is an immutable map.

Reply:
Does changing to `<M: MutableMap<K, V>>` work?

Reply:
I think having type-safety is not enforcing mutability/immutability of the structure you are serializing/de-serializing. If `M: MutableMap<K, V>>` this means that `M` is a subtype of `Map<K, V>` so you could maybe use that if you want.

Reply:
Okay so let me clarify my intentions first.

Suppose we need an externalizer for `val data: Map<String, String>`. My goal is for the usage to look like this:

```
val x: Map<String, String> = MapExternalizer(StringExternalizer, StringExternalizer).read(input) // compiles okay
val y: MutableMap<String, String> = MapExternalizer(StringExternalizer, StringExternalizer).read(input) // fails to compile
```

That is, the `read` method should return an immutable map.

The current implementation of `MapExternalizer` achieves that goal (even though I agree that the unsafe casting is not great).

I tried your suggestions, but it required removing the default value of `newMap`:
```
open class MapExternalizer<K, V, M : MutableMap<K, V>>(
    ...
    private val newMap: (size: Int) -> M // Removed: MutableMap<K, V> = { size -> LinkedHashMap(size) }
}    
```    

So the usage will look like this:
```
val x: Map<String, String> = MapExternalizer(StringExternalizer, StringExternalizer, {LinkedHashMap()}).read(input) // compiles okay
val y: MutableMap<String, String> = MapExternalizer(StringExternalizer, StringExternalizer, {LinkedHashMap()}).read(input) // compiles okay
```

There are 2 effects at the usage:
  - I have to provide the extra `{LinkedHashMap()}` argument, which should be an internal detail of `MapExternalizer`.
  - The `read` method returns a mutable map now.
  
If there is an alternative implementation that achieves my previous goal, that would be great.

Reply:
Is `LinkedHashMapExternalizer` the only implementation? If so there is no need for this base class.

Also, please note that you can solve this issue by creating parametrized subclasses of `MapExternalizer` with concrete `newMap` params, as you've done in `LinkedHashMapExternalizer`.

Reply:
At one point during implementation, I had this two classes:
```
class ClasspathEntrySnapshot(
    val classSnapshots: **LinkedHashMap**<String, ClassSnapshot>
)

class PackageFacadeKotlinClassSnapshot(
    val packageMembers: **Map**<FqName, String>
) : KotlinClassSnapshot()
```

So I created `MapExternalizer` and let `LinkedHashMapExternalizer` extend it.

But later on, I changed `PackageFacadeKotlinClassSnapshot` to:
```
class PackageFacadeKotlinClassSnapshot(
    val packageMemberNames: **Set**<String>
) : KotlinClassSnapshot()
```

So `MapExternalizer` is unused now (except by `LinkedHashMapExternalizer`). I decided to keep it because:
  - The use of `Map` is common, and we'll have `MapExternalizer` already available when we need it.
  - `ListExternalizer` and `SetExternalizer` extend `GenericCollectionExternalizer` (which btw has the same unsafe typecast issue), so if we have `LinkedHashMapExternalizer` extend `MapExternalizer`, it will be the same pattern.
  
That said, I do agree `MapExternalizer` is currently unused so I can also remove it. (The only issue is that if I have `val data: Map<X, Y>` in the future, I will need to introduce `MapExternalizer` back again -- which is fine.)

WDYT?

Reply:
Removing it, and having just a single implementation, sounds good to me.

Reply:
Sure thing, I'll fix that.